### PR TITLE
drivers: spi: nrfx_spim: Add support for device runtime PM

### PR DESCRIPTION
**Pull request by adamkondraciuk**
Enable the device runtime power management on the SPIM shim.

### CODE REVIEW

**Review by kl-cruz — DISMISSED** [review]

File: tests/drivers/spi/spi_controller_peripheral/boards/nrf52840dk_nrf52840.overlay [hunk 1]
```diff
@@ -53,6 +53,7 @@
 	dut_spi_dt: test-spi-dev@0 {
 		compatible = "vnd,spi-device";
 		reg = <0>;
+		spi-max-frequency = <DT_FREQ_M(8)>;
```
Comment (anangl) [inline tests/drivers/spi/spi_controller_peripheral/boards/nrf52840dk_nrf52840.overlay:56]: Why?

**Review by anangl — CHANGES_REQUESTED** [review]

File: tests/drivers/spi/spi_controller_peripheral/boards/nrf52840dk_nrf52840.overlay (hunk 1 above)
Comment (adamkondraciuk) [reply tests/drivers/spi/spi_controller_peripheral/boards/nrf52840dk_nrf52840.overlay:57]: because:
`devicetree error: 'spi-max-frequency' is marked as required in 'properties:' in /home/adko/zephyrproject/zephyr/dts/bindings/test/vnd,spi-device.yaml, but does not appear in <Node /soc/spi@4002f000/test-spi-dev@0 in '/home/adko/zephyrproject/zephyr/misc/empty_file.c'>`

**Review by kl-cruz — DISMISSED** [review]

**Review by nika-nordic — DISMISSED** [review]

**Comment by rda-emcraft** [comment]
This code works incorrectly in the case of  `SPI_HOLD_ON_CS` or/and  `SPI_LOCK_ON`.

Here's what happens when SPI transactions are split into two calls of spi_transceive()
```
[00:24:15.862,182] <wrn> pm_device: Unbalanced suspend
[00:24:15.863,128] <wrn> pm_device: Unbalanced suspend
[00:24:15.966,217] <wrn> pm_device: Unbalanced suspend
[00:24:15.966,857] <wrn> pm_device: Unbalanced suspend
[00:24:16.816,741] <wrn> pm_device: Unbalanced suspend
```
And I've got lucky because switching to power down does not break the transaction. Probably idle pin states between the bytes match power down states.

It does not make sense to call `pm_device_runtime_put(dev)` if SPI lock is not released or if CS hold is requested.
Similarly it does not make sense to call `pm_device_runtime_get(dev)` if SPI lock or CS is held.

By the way, all zephyr SPI drivers with device runtime support I've seen does not handle this flags. Probably generic solution will be to integrate device runtime calls into spi_context.h.

**Comment by nordic-piks** [comment]
> This code works incorrectly in the case of `SPI_HOLD_ON_CS` or/and `SPI_LOCK_ON`.
> 
> Here's what happens when SPI transactions are split into two calls of spi_transceive()
> 
> ```
> [00:24:15.862,182] <wrn> pm_device: Unbalanced suspend
> [00:24:15.863,128] <wrn> pm_device: Unbalanced suspend
> [00:24:15.966,217] <wrn> pm_device: Unbalanced suspend
> [00:24:15.966,857] <wrn> pm_device: Unbalanced suspend
> [00:24:16.816,741] <wrn> pm_device: Unbalanced suspend
> ```
> 
> And I've got lucky because switching to power down does not break the transaction. Probably idle pin states between the bytes match power down states.
> 
> It does not make sense to call `pm_device_runtime_put(dev)` if SPI lock is not released or if CS hold is requested. Similarly it does not make sense to call `pm_device_runtime_get(dev)` if SPI lock or CS is held.
> 
> By the way, all zephyr SPI drivers with device runtime support I've seen does not handle this flags. Probably generic solution will be to integrate device runtime calls into spi_context.h.

@rda-emcraft Can you rise an issue with this please? That will make easier to implement this.

**Comment by nordic-piks** [comment]
Hm, maybe issue is not needed as this is PR only.

**Comment by rda-emcraft** [comment]
> 
> @rda-emcraft Can you rise an issue with this please? That will make easier to implement this.
@nordic-piks, @adamkondraciuk 
I may raise issue for spi_ambiq.c, spi_ll_stm32.c and spi_smartbond.c that already accepted into the tree and has the same deficiency. 

I have read more of spi_context.h and now I think you should call `pm_device_runtime_get(dev)` along with locking `ctx->lock` and `pm_device_runtime_put(dev)` when unlocking `ctx->lock`. Implementing this might require passing device pointer to `spi_context_{lock,release,complete,release,unlock_unconditionally}()`. Probably will have this implemented later today.

What I'm trying to say is that part of runtime device PM (_get() and put()) for SPI drivers can be implemented in generic SPI context handling code. Driver only need to supply initialization and code to do real PM operations.

Anyway, I'm really new in Zephyr, so any guidance is appreciated.

And I'm sitting on Zephyr 3.5.99 and calling `pm_device_runtime_put(dev)` from IRQ context is a no go here. I'm replacing it with `pm_device_runtime_put_async(dev)`.

**Comment by rda-emcraft** [comment]
@adamkondraciuk 

I've created pull request to your branch implementing proposed changes.

File: drivers/spi/spi_nrfx_spim.c [hunk 2]
```diff
@@ -568,21 +575,20 @@ static const struct spi_driver_api spi_nrfx_driver_api = {
...
 static int spim_nrfx_pm_action(const struct device *dev,
 			       enum pm_device_action action)
 {
-	int ret = 0;
+	int ret = -ENOTSUP;
 	struct spi_nrfx_data *dev_data = dev->data;
 	const struct spi_nrfx_config *dev_config = dev->config;
 
 	switch (action) {
+	case PM_DEVICE_ACTION_TURN_ON:
```
Comment (nordic-krch) [inline drivers/spi/spi_nrfx_spim.c:586]: it's not needed any more. Function can return `-ENOTSUP` in that case and it's accepted by the PM.

File: drivers/spi/spi_nrfx_spim.c [hunk 3]
```diff
@@ -596,19 +602,14 @@ static int spim_nrfx_pm_action(const struct device *dev,
 
 		ret = pinctrl_apply_state(dev_config->pcfg,
```
Comment (nordic-krch) [inline drivers/spi/spi_nrfx_spim.c:603]: suspend action shall be wrapped by `IS_ENABLED(CONFIG_PM_DEVICE)` and it cannot be called when PM is not enabled (sleep state in pinctrl does not exist, too).
Maybe similar approach could be used like here? https://github.com/zephyrproject-rtos/zephyr/blob/afce5329f772a4306af699c7a06efddf93cadca0/drivers/pwm/pwm_nrfx.c#L302 (already used in uart_nrfx_uarte.c as well).

Comment (rda-emcraft) [reply drivers/spi/spi_nrfx_spim.c:603]: Can do. And we also need to start suspended with `PINCTRL_STATE_SLEEP` set.

Patch is updated.

Actually, both places are not the part I expected comments about. I'm more worried about changes introduced to spi_context.h and the whole idea that device runtime PM should keep device available while SPI context lock is held.

Also, if this is indeed correct solution, other SPI drivers (spi_ambiq.c, spi_ll_stm32.c and spi_smartbond.c) need to be updated to use same approach. Likely this is best to be handled in a separate issue.

File: drivers/spi/spi_nrfx_spim.c [hunk 4]
```diff
@@ -90,6 +91,9 @@ static inline void finalize_spi_transaction(const struct device *dev, bool deact
 	if (NRF_SPIM_IS_320MHZ_SPIM(reg) && !(dev_data->ctx.config->operation & SPI_HOLD_ON_CS)) {
 		nrfy_spim_disable(reg);
 	}
+	if (IS_ENABLED(CONFIG_PM_DEVICE_RUNTIME)) {
```
Comment (anangl) [inline drivers/spi/spi_nrfx_spim.c:94]: This is unneeded. `pm_device_runtime_put()` can be called without enabled `PM_DEVICE_RUNTIME`. Same goes for `pm_device_runtime_get()`.

File: drivers/spi/spi_nrfx_spim.c [hunk 5]
```diff
@@ -571,21 +578,20 @@ static DEVICE_API(spi, spi_nrfx_driver_api) = {
...
 	switch (action) {
+	case PM_DEVICE_ACTION_TURN_ON:
+		ret = 0;
+		break;
 	case PM_DEVICE_ACTION_RESUME:
 		ret = pinctrl_apply_state(dev_config->pcfg,
 					  PINCTRL_STATE_DEFAULT);
-		if (ret < 0) {
-			return ret;
-		}
```
Comment (anangl) [inline drivers/spi/spi_nrfx_spim.c:588]: Why is this removed (also in `PM_DEVICE_ACTION_SUSPEND`)?

File: tests/drivers/spi/spi_controller_peripheral/boards/nrf52840dk_nrf52840.overlay [hunk 6]
```diff
@@ -50,10 +50,12 @@
 	pinctrl-names = "default", "sleep";
 	overrun-character = <0x00>;
 	cs-gpios = <&gpio1 11 GPIO_ACTIVE_LOW>;
+	zephyr,pm-device-runtime-auto;
 	dut_spi_dt: test-spi-dev@0 {
 		compatible = "vnd,spi-device";
 		spi-max-frequency = <DT_FREQ_M(8)>;
 		reg = <0>;
+		spi-max-frequency = <DT_FREQ_M(8)>;
```
Comment (anangl) [inline tests/drivers/spi/spi_controller_peripheral/boards/nrf52840dk_nrf52840.overlay:58]: No need to duplicate this property.

File: drivers/spi/spi_nrfx_spim.c (hunk 2 above)
Comment (nika-nordic) [reply drivers/spi/spi_nrfx_spim.c]: aligned to scheme used in PWM and UARTE drivers

File: drivers/spi/spi_nrfx_spim.c (hunk 3 above)
Comment (nika-nordic) [reply drivers/spi/spi_nrfx_spim.c]: aligned to scheme used in PWM and UARTE drivers

File: drivers/spi/spi_nrfx_spim.c (hunk 4 above)
Comment (nika-nordic) [reply drivers/spi/spi_nrfx_spim.c]: removed redundant conditionals

File: drivers/spi/spi_nrfx_spim.c (hunk 5 above)
Comment (nika-nordic) [reply drivers/spi/spi_nrfx_spim.c:588]: it is aligned to how UARTE and PWM drivers behave in this manner - they ignore return code:
-  https://github.com/zephyrproject-rtos/zephyr/blob/fe0e2dbc6044ffa74b27bfcba335980bc5a6ae64/drivers/serial/uart_nrfx_uarte.c#L2193
- https://github.com/zephyrproject-rtos/zephyr/blob/fe0e2dbc6044ffa74b27bfcba335980bc5a6ae64/drivers/serial/uart_nrfx_uarte.c#L2265
- https://github.com/zephyrproject-rtos/zephyr/blob/fe0e2dbc6044ffa74b27bfcba335980bc5a6ae64/drivers/pwm/pwm_nrfx.c#L278
- https://github.com/zephyrproject-rtos/zephyr/blob/fe0e2dbc6044ffa74b27bfcba335980bc5a6ae64/drivers/pwm/pwm_nrfx.c#L317

File: tests/drivers/spi/spi_controller_peripheral/boards/nrf52840dk_nrf52840.overlay (hunk 6 above)
Comment (nika-nordic) [reply tests/drivers/spi/spi_controller_peripheral/boards/nrf52840dk_nrf52840.overlay]: removed duplicated entry

**Review by anangl — APPROVED** [review]

**Review by nordic-segl — APPROVED** [review]

**Review by masz-nordic — APPROVED** [review]
